src/azul/indexer/__init__.py

```diff
@@ -83,10 +83,43 @@ class SimpleSourceSpec(SourceSpec['SimpleSourceSpec']):
 
     @classmethod
     def parse(cls, spec: str) -> 'SimpleSourceSpec':
-        return cls(name=spec)
+        """
+        >>> s = SimpleSourceSpec.parse('https://foo.edu:12')
+        >>> s
+        SimpleSourceSpec(prefix='12', name='https://foo.edu')
+
+        >>> s = SimpleSourceSpec.parse('foo:bar/baz:12')
+        >>> s
+        SimpleSourceSpec(prefix='12', name='foo:bar/baz')
+
+        >>> s = SimpleSourceSpec.parse('foo')
+        >>> s
+        SimpleSourceSpec(prefix='', name='foo')
+        """
+        # FIXME: Move parsing of prefix to SourceSpec
+        #        https://github.com/DataBiosphere/azul/issues/3073
+        if ':' in spec:
+            name, _, prefix = spec.rpartition(':')
+        else:
+            name = spec
+            prefix = ''
+        return cls(prefix=prefix, name=name)
 
     def __str__(self) -> str:
-        return self.name
+        """
+        >>> s = SimpleSourceSpec(prefix='12', name='foo:bar/baz')
+        >>> str(s)
+        'foo:bar/baz:12'
+
+        >>> s = SimpleSourceSpec(prefix=':12', name='foo:bar/baz')
+        >>> str(s)
+        Traceback (most recent call last):
+        ...
+        AssertionError
+
+        """
+        assert ':' not in self.prefix
```

Comment:
Same assert wouldn't hurt in `parse`

Reply:
I think doing some type of UUID validation would be good to prevent stuff like;
```
SimpleSourceSpec.parse('https://amar.edu')
SimpleSourceSpec(prefix='//amar.edu', name='https')
```

Reply:
As an alternative to both assertions; why not add a attrs validator to the prefix attribute?

```
@attr.s(frozen=True, auto_attribs=True, kw_only=True)
class SourceSpec(ABC, Generic[SOURCE_SPEC]):

    prefix: str = attr.ib(default='',
                          validator=validate_uuid)
```

We are ensured our prefixes are valid when the objects are instantiated, and we can remove the assertions in both locations, as it would fail due to the prefix being invalid w/ a `:` present.

Reply:
Assertions and validations are different things. The former express program invariants, the latter express constraints on program input. We'd still want the assertion, even if we did a validation. I would write a `__post_init` (less code that a validator) and in it, require that the prefix is a valid UUID prefix and then assert that it doesn't have a colon in it.